### PR TITLE
Fix handling memory of max size on 32-bit systems: change hard memory limit

### PR DESCRIPTION
**Pull request by gumb0**
To support 32-bit systems I think it's better not to add another check for `size_t` overflow, but instead rely on existing check against hard memory limit and cap this limit to below 4 Gb (e.g. to 65535 pages) for 32-bit systems.

This way the logic is simpler in `allocate_memory` and `grow_memory` (I leave onle assertions for `size_t` overflow in these places), there's no additional check to test, the only complication is that some unit tests have different expectations for 32-bit build.
(Also `(module (memory 0 65536))` spec test won't pass on 32-bit)

Hard memory limit could be thought of as maximum possible memory size that system supports, then for 32-bit it shoud be below 4 Gb.

### CODE REVIEW

File: lib/fizzy/limits.hpp [hunk 1]
```diff
@@ -4,17 +4,30 @@
...
 #include <cstdint>
+#include <limits>
 
 namespace fizzy
 {
 /// The page size as defined by the WebAssembly 1.0 specification.
 constexpr uint32_t PageSize = 65536;
 
+/// Convert memory size in pages to size in bytes.
+inline constexpr uint64_t memory_pages_to_bytes(uint32_t pages) noexcept
```
Comment (gumb0) [inline lib/fizzy/limits.hpp:18]: Any better place for this helper?

File: lib/fizzy/limits.hpp [hunk 2]
```diff
@@ -4,17 +4,31 @@
...
+{
+    return uint64_t{pages} * PageSize;
+}
+
 /// The maximum memory page limit as defined by the specification.
```
Comment (gumb0) [inline lib/fizzy/limits.hpp:23]: Maybe these comments could be improved.

**Review by chfast — APPROVED** [review]
Looks good to be although I'm not sure if having the same limit for all architectures wouldn't be cleaner. Having the `allocate()` helper could check if the requested memory size would fit into `size_t`. Otherwise the allocation should fail imminently.

**Comment by gumb0** [comment]
> Looks good to be although I'm not sure if having the same limit for all architectures wouldn't be cleaner. Having the `allocate()` helper could check if the requested memory size would fit into `size_t`. Otherwise the allocation should fail imminently.

Yeah I think it would be possible, the 2nd helper will be called in 2 places: `allocate_memory` and `grow_memory` where assertions currently are. 

But in `allocate_memory` if the check looks like `if (!valid_allocation(size)) throw(...)` throwing lines cannot be covered in 64-bit build, unless we put the throw into the 3d helper...

**Comment by chfast** [comment]
> But in `allocate_memory` if the check looks like `if (!valid_allocation(size)) throw(...)` throwing lines cannot be covered in 64-bit build, unless we put the throw into the 3d helper...

This can be in `if constexpr` section. And we can also collect coverage for 32-bit builds. If these are the only problems, I'd try to do it.

**Comment by gumb0** [comment]
Closing in favor of #808

**Pull request closed without merging**